### PR TITLE
Hive: Use HiveClientPool cache instead of HiveCatalog global cache.

### PR DESCRIPTION
**Pull request by lcspinter**
HiveClientPool instances should be shared between HiveCatalogs if they are targeting the same HiveMetastore. 

What I have done
- Removed HiveCatalog global cache calls and marked HiveCatalogs#loadCatalog() as deprecated
- Created cache inside CachedClientPoo
- The cache is configured to remove HiveClientPool entries if they were not used for a certain amount of time (client.pool.cache.eviction-interval-ms). This makes sure we are cleaning up every unused HiveClientPool connection. 
- Removed the finalize() method implementation from the HiveCatalog. Previously the cleanup of the HiveClientPool connections was handled when the HiveCatalog was GC'd. 
- Instead of passing the HiveClientPool to the HiveTableOperations, now the HiveClientPoolProvider is passed to fetch the HiveClientPool instance from the cache.
- HiveCatalog is not an instance of Closable anymore. 
- Created unit test to validate cache eviction.

### CODE REVIEW

**Comment by lcspinter** [comment]
@rymurr @pvary @marton-bod Could you please review this PR? Thank you

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 1]
```diff
@@ -51,36 +55,48 @@
...
+public class HiveCatalog extends BaseMetastoreCatalog implements SupportsNamespaces, Configurable {
   private static final Logger LOG = LoggerFactory.getLogger(HiveCatalog.class);
 
+  private static final String CACHE_CLEANER_INTERVAL = "iceberg.hive.client-pool-cache-cleaner-interval";
+  private static final long CACHE_CLEANER_INTERVAL_DEFAULT = TimeUnit.SECONDS.toMillis(30);
+  private static final String CACHE_EVICTION_INTERVAL = "iceberg.hive.client-pool-cache-eviction-interval";
+  private static final long CACHE_EVICTION_INTERVAL_DEFAULT = TimeUnit.MINUTES.toMillis(5);
+
+  @VisibleForTesting
+  static final Cache<String, Pair<HiveClientPool, Long>> CLIENT_POOL_CACHE = Caffeine.newBuilder()
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:79]: Just a question:
Wouldn't it make sense to put this into the HiveClientPool instead with all the scaffolding.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 2]
```diff
@@ -604,4 +596,34 @@ public void setConf(Configuration conf) {
...
+      synchronized (HiveCatalog.class) {
+        if (cleaner == null) {
+          cleaner = Executors.newSingleThreadScheduledExecutor(
+                  new ThreadFactoryBuilder()
+                          .setDaemon(true)
+                          .setNameFormat("iceberg-client-pool-cache-cleaner-%d")
+                          .build());
+        }
+        ScheduledFuture<?> futures = cleaner.scheduleWithFixedDelay(() -> CLIENT_POOL_CACHE.asMap().entrySet().stream()
+                .filter(e -> e.getValue().second() + evictionInterval <= System.currentTimeMillis())
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:620]: Maybe put the eviction time instead of the pair, and calculate the currentTimeMillis outside of the stream to shave off a few cycles.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 3]
```diff
@@ -604,4 +596,34 @@ public void setConf(Configuration conf) {
...
+                .forEach(e -> {
+                  HiveClientPool pool = e.getValue().first();
+                  CLIENT_POOL_CACHE.invalidate(e.getKey());
+                  pool.close();
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:624]: Could we have some problem with concurrency?
We decide to evict, but just after the check someone gets this pool, but we close it...

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 4]
```diff
@@ -604,4 +596,34 @@ public void setConf(Configuration conf) {
   public Configuration getConf() {
     return conf;
   }
+
+  @VisibleForTesting
+  HiveClientPool loadHiveClientPool(Configuration configuration) {
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:601]: Could we just call it `clientPool`? I would prefer not to "leak" information about how it is handled internally

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java [hunk 5]
```diff
@@ -400,6 +401,20 @@ public void testTableName() {
     }
   }
 
+  @Test
+  public void testClientPoolCleaner() throws InterruptedException {
+    HiveCatalog catalog = new HiveCatalog(hiveConf);
+    HiveClientPool clientPool1 = catalog.loadHiveClientPool(hiveConf);
+    Thread.sleep(TimeUnit.SECONDS.toMillis(5));
```
Comment (pvary) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java:408]: I have found that `Clock` is very useful when we run tests against current time and do not want to use Thread.sleeps...

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java [hunk 6]
```diff
@@ -215,6 +216,8 @@ private void initConf(HiveConf conf, int port) {
     conf.set(HiveConf.ConfVars.METASTORE_TRY_DIRECT_SQL.varname, "false");
     conf.set(HiveConf.ConfVars.METASTORE_DISALLOW_INCOMPATIBLE_COL_TYPE_CHANGES.varname, "false");
     conf.set("iceberg.hive.client-pool-size", "2");
+    conf.setLong("iceberg.hive.client-pool-cache-cleaner-interval", TimeUnit.SECONDS.toMillis(5));
```
Comment (pvary) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java:219]: Could we use the consts here?

File: spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java [hunk 7]
```diff
@@ -73,9 +73,6 @@ public static void startMetastoreAndSpark() {
 
   @AfterClass
   public static void stopMetastoreAndSpark() {
-    if (catalog != null) {
-      catalog.close();
-    }
```
Comment (pvary) [inline spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java:78]: Should we make this consistent with other tests and set the catalog to null?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 2 above)
Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:620]: Good idea. I changed the code accordingly.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 3 above)
Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:624]: You are right, I missed this one. We should synchronize on the CLIENT_POOL_CACHE when inserting/deleting entries.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 4 above)
Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:601]: Fixed it.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java (hunk 5 above)
Comment (lcspinter) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java:408]: I choose Thread.sleep() because we are using this pattern in other test suits as well. Anyway, I changed this part to TimeUnit.SECONDS.sleep() for better readability.

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java (hunk 6 above)
Comment (lcspinter) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java:219]: Done.

File: spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java (hunk 7 above)
Comment (lcspinter) [reply spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java:78]: It is nulled out in the next line.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 1 above)
Comment (rymurr) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:79]: I agree. From my reading of this PR most of the client pool related stuff is a separate concern to the catalog itself and should be in a separate class.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 8]
```diff
@@ -98,15 +97,16 @@
   private final long lockCheckMinWaitTime;
   private final long lockCheckMaxWaitTime;
   private final FileIO fileIO;
+  private final HiveCatalog catalog;
 
-  protected HiveTableOperations(Configuration conf, HiveClientPool metaClients, FileIO fileIO,
-                                String catalogName, String database, String table) {
+  protected HiveTableOperations(Configuration conf, FileIO fileIO, HiveCatalog catalog,
```
Comment (rymurr) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:102]: Along w/ my comment above I don't think the table ops should get to see the entire catalog, rather just the client pool

**Review by rymurr — COMMENTED** [review]
Love this @lcspinter ive always wanted to remove the finalize method! My main comment would be to isolate the client pool from the catalog.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 9]
```diff
@@ -50,25 +55,37 @@
...
 import org.slf4j.LoggerFactory;
 
-public class HiveCatalog extends BaseMetastoreCatalog implements Closeable, SupportsNamespaces, Configurable {
+public class HiveCatalog extends BaseMetastoreCatalog implements SupportsNamespaces, Configurable {
   private static final Logger LOG = LoggerFactory.getLogger(HiveCatalog.class);
 
+  public static final String CACHE_CLEANER_INTERVAL = "iceberg.hive.client-pool-cache-cleaner-interval";
+  private static final long CACHE_CLEANER_INTERVAL_DEFAULT = TimeUnit.SECONDS.toMillis(30);
+  public static final String CACHE_EVICTION_INTERVAL = "iceberg.hive.client-pool-cache-eviction-interval";
+  private static final long CACHE_EVICTION_INTERVAL_DEFAULT = TimeUnit.MINUTES.toMillis(5);
```
Comment (marton-bod) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:76]: Quick question: is there a rationale behind the 5 minutes cache eviction as a default? Does it typically become stale somehow after a few minutes that we need to replace it? Just wondering, given that the pool recreation has a cost

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 10]
```diff
@@ -547,4 +541,42 @@ public void setConf(Configuration conf) {
   public Configuration getConf() {
     return conf;
   }
+
+  @VisibleForTesting
+  HiveClientPool clientPool() {
+    synchronized (CLIENT_POOL_CACHE) {
+      String metastoreUri = conf.get(HiveConf.ConfVars.METASTOREURIS.varname, "");
+      Pair<HiveClientPool, Long> cacheEntry = CLIENT_POOL_CACHE.getIfPresent(metastoreUri);
+      HiveClientPool clientPool = cacheEntry == null ? new HiveClientPool(clientPoolSize, conf) : cacheEntry.first();
```
Comment (marton-bod) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:550]: Can we restrict the scope of the synchronised block to contain only the cache population part? that way we could decrease the contention on the more frequent cache lookup operation. If `cacheEntry` is null, we enter the block, and re-check if it's been populated by another thread in the meantime

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 11]
```diff
@@ -547,4 +541,42 @@ public void setConf(Configuration conf) {
...
+        long cleanerInterval = conf.getLong(CACHE_CLEANER_INTERVAL, CACHE_CLEANER_INTERVAL_DEFAULT);
+        ScheduledFuture<?> futures = cleaner.scheduleWithFixedDelay(() -> {
+          synchronized (CLIENT_POOL_CACHE) {
+            long currentTime = System.currentTimeMillis();
+            CLIENT_POOL_CACHE.asMap().entrySet().stream()
+                    .filter(e -> e.getValue().second() <= currentTime)
+                    .forEach(e -> {
+                      HiveClientPool pool = e.getValue().first();
+                      CLIENT_POOL_CACHE.invalidate(e.getKey());
+                      pool.close();
```
Comment (marton-bod) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:575]: If we decide to close the pool here regardless of its usage by any clients, it's probably worth pointing out in documentation that the pool has an expiration date and users should not cache it in a variable for too long. For example:
```
pool = getPool()
// some long running other operation
pool.getTable("table"); // -> FAIL
```

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 12]
```diff
@@ -50,25 +55,37 @@
...
 import org.apache.iceberg.util.PropertyUtil;
 import org.apache.thrift.TException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class HiveCatalog extends BaseMetastoreCatalog implements Closeable, SupportsNamespaces, Configurable {
+public class HiveCatalog extends BaseMetastoreCatalog implements SupportsNamespaces, Configurable {
   private static final Logger LOG = LoggerFactory.getLogger(HiveCatalog.class);
 
+  public static final String CACHE_CLEANER_INTERVAL = "iceberg.hive.client-pool-cache-cleaner-interval";
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:73]: Where should these be set? We avoid setting things in Hadoop `Configuration`.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 13]
```diff
@@ -19,11 +19,16 @@
 
 package org.apache.iceberg.hive;
 
-import java.io.Closeable;
-import java.util.Arrays;
+
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:22]: Nit: please remove unnecessary whitespace changes.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 14]
```diff
@@ -547,4 +541,42 @@ public void setConf(Configuration conf) {
...
     return conf;
   }
+
+  @VisibleForTesting
+  HiveClientPool clientPool() {
+    synchronized (CLIENT_POOL_CACHE) {
+      String metastoreUri = conf.get(HiveConf.ConfVars.METASTOREURIS.varname, "");
+      Pair<HiveClientPool, Long> cacheEntry = CLIENT_POOL_CACHE.getIfPresent(metastoreUri);
+      HiveClientPool clientPool = cacheEntry == null ? new HiveClientPool(clientPoolSize, conf) : cacheEntry.first();
+      CLIENT_POOL_CACHE.put(metastoreUri, Pair.of(clientPool, System.currentTimeMillis() + evictionInterval));
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:551]: Why use a cache but not delegate thread safety to it? Shouldn't this use `get` with an init function?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 15]
```diff
@@ -547,4 +541,42 @@ public void setConf(Configuration conf) {
...
+                          .setDaemon(true)
+                          .setNameFormat("iceberg-client-pool-cache-cleaner-%d")
+                          .build());
+        }
+        long cleanerInterval = conf.getLong(CACHE_CLEANER_INTERVAL, CACHE_CLEANER_INTERVAL_DEFAULT);
+        ScheduledFuture<?> futures = cleaner.scheduleWithFixedDelay(() -> {
+          synchronized (CLIENT_POOL_CACHE) {
+            long currentTime = System.currentTimeMillis();
+            CLIENT_POOL_CACHE.asMap().entrySet().stream()
+                    .filter(e -> e.getValue().second() <= currentTime)
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:571]: Can't expiration be delegated to the cache?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 16]
```diff
@@ -547,4 +541,42 @@ public void setConf(Configuration conf) {
...
+    synchronized (CLIENT_POOL_CACHE) {
+      String metastoreUri = conf.get(HiveConf.ConfVars.METASTOREURIS.varname, "");
+      Pair<HiveClientPool, Long> cacheEntry = CLIENT_POOL_CACHE.getIfPresent(metastoreUri);
+      HiveClientPool clientPool = cacheEntry == null ? new HiveClientPool(clientPoolSize, conf) : cacheEntry.first();
+      CLIENT_POOL_CACHE.put(metastoreUri, Pair.of(clientPool, System.currentTimeMillis() + evictionInterval));
+      return clientPool;
+    }
+  }
+
+  private void scheduleCacheCleaner() {
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:556]: I think it would be cleaner to always keep the pool around and never close it until the cache is cleaned up. Then the pool would always be shared and could be kept as an instance variable inside the catalog. It would also make cleanup less coarse. We could expire individual connections that are unused instead of expiring whole pools. That aligns better with usage patterns and keeps the complexity inside the pool rather than in the cache.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 8 above)
Comment (rdblue) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:102]: Agreed.

Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: @rymurr @rdblue You are right. I changed this part.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 10 above)
Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java]: @marton-bod Thanks for the review. I delegated the cache update/cleanup to the cache itself, so now it became threadsafe.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 9 above)
Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java]: @marton-bod There is no particular reason for choosing this value.
If we keep the ClientPool and its connections open we might end up with a memory leak. (With the removal of the finalize() method nobody is closing them). To tackle this issue, I decided to close every unused ClientPool after a certain amount of time. This threshold is configurable and should be updated based on the load/usage of the cluster.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 12 above)
Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java]: Right, moved these configs to `CatalogProperties`.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 13 above)
Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java]: Done

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 14 above)
Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java]: Good idea. I delegated the complete update/invalidate to the cache.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 15 above)
Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java]: Done.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveClientPoolProvider.java [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.util.concurrent.TimeUnit;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.iceberg.CatalogProperties;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+
+public class HiveClientPoolProvider {
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveClientPoolProvider.java:30]: Could this object implement the `ClientPool` interface and maybe renamed to `CachedClientPool`?

This way we can push this `ClientPool` to the HiveTableOperations and do not need to change there

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 18]
```diff
@@ -102,14 +94,17 @@ public void initialize(String inputName, Map<String, String> properties) {
...
-    int clientPoolSize = PropertyUtil.propertyAsInt(properties,
-        CatalogProperties.CLIENT_POOL_SIZE, CatalogProperties.CLIENT_POOL_SIZE_DEFAULT);
-    this.clients = new HiveClientPool(clientPoolSize, this.conf);
-    this.createStack = Thread.currentThread().getStackTrace();
-    this.closed = false;
+    this.conf.setInt(CatalogProperties.CLIENT_POOL_SIZE, PropertyUtil.propertyAsInt(properties,
+            CatalogProperties.CLIENT_POOL_SIZE, CatalogProperties.CLIENT_POOL_SIZE_DEFAULT));
+
+    this.conf.setLong(CatalogProperties.CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS,
+            PropertyUtil.propertyAsLong(properties, CatalogProperties.CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS,
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:101]: Is there a particular reason to not push this logic to the `HiveClientPoolProvider` and add a property map as an input parameter?
Or otherwise specifically providing it as a constructor parameter?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveClientPoolProvider.java (hunk 17 above)
Comment (pvary) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveClientPoolProvider.java:30]: Maybe even something like this:
https://github.com/apache/iceberg/blob/94705c6799f3b283b45e47e6f9b46b5d081ed666/core/src/main/java/org/apache/iceberg/CachingCatalog.java#L35-L37

Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveClientPoolProvider.java]: Refactored it.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 18 above)
Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java]: Done.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 19]
```diff
@@ -83,12 +78,15 @@ public HiveCatalog() {
...
-    this.createStack = Thread.currentThread().getStackTrace();
-    this.closed = false;
     this.fileIO = new HadoopFileIO(conf);
+    Map<String, String> properties = ImmutableMap.of(
+            CatalogProperties.CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS,
+            conf.get(CatalogProperties.CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS),
+            CatalogProperties.CLIENT_POOL_SIZE,
+            conf.get(CatalogProperties.CLIENT_POOL_SIZE)
+    );
+    this.clientPool = new CachedClientPool(conf, properties);
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:89]: Could we rename this back to `clients`? So there is fewer changes...

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 20]
```diff
@@ -100,8 +99,9 @@
   private final long lockCheckMinWaitTime;
   private final long lockCheckMaxWaitTime;
   private final FileIO fileIO;
+  private final CachedClientPool metaClients;
 
-  protected HiveTableOperations(Configuration conf, HiveClientPool metaClients, FileIO fileIO,
+  protected HiveTableOperations(Configuration conf, CachedClientPool metaClients, FileIO fileIO,
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:104]: What is the least restrictive parameter that we can use here? `HiveClientPool`, or `ClientPool`?

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommitLocks.java [hunk 21]
```diff
@@ -99,7 +105,7 @@ public void before() throws Exception {
 
     Assert.assertEquals(2, ops.current().schema().columns().size());
 
-    spyOps = spy(new HiveTableOperations(overriddenHiveConf, spyClientPool, ops.io(), catalog.name(),
+    spyOps = spy(new HiveTableOperations(overriddenHiveConf,  spyCachedClientPool, ops.io(), catalog.name(),
```
Comment (pvary) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCommitLocks.java:108]: nit: extra space

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 20 above)
Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: I changed it to ClientPool

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveClientPool.java [hunk 22]
```diff
@@ -44,7 +44,7 @@ public void testConf() {
     conf.set(HiveConf.ConfVars.METASTOREWAREHOUSE.varname, "file:/mywarehouse/");
     conf.setInt("iceberg.hive.client-pool-size", 10);
 
-    HiveClientPool clientPool = new HiveClientPool(conf);
+    HiveClientPool clientPool = new HiveClientPool(10, conf);
```
Comment (pvary) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveClientPool.java:47]: This one is not needed anymore

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+  @Test
+  public void testClientPoolCleaner() throws InterruptedException {
+    String metastoreUri = hiveConf.get(HiveConf.ConfVars.METASTOREURIS.varname, "");
+    CachedClientPool clientPool = new CachedClientPool(hiveConf, Collections.emptyMap());
+    HiveClientPool clientPool1 = clientPool.clientPool();
+    Assert.assertTrue(CachedClientPool.clientPoolCache().getIfPresent(metastoreUri) == clientPool1);
+    TimeUnit.SECONDS.sleep(8);
+    HiveClientPool clientPool2 = clientPool.clientPool();
+    Assert.assertTrue(clientPool1 == clientPool2);
+    TimeUnit.SECONDS.sleep(15);
```
Comment (marton-bod) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java:39]: Shall we reference the 10 second eviction interval here from the base class? That way we could replace the magic numbers with stuff like `HiveMetastoreTest.EVICTION_INTERVAL - TimeUnit.Seconds(5).toMillis()` and `HiveMetastoreTest.EVICTION_INTERVAL + TimeUnit.Seconds(5).toMillis()`. What do you think?

File: mr/src/main/java/org/apache/iceberg/mr/Catalogs.java [hunk 24]
```diff
@@ -212,7 +214,7 @@ public static boolean hiveCatalog(Configuration conf) {
           LOG.info("Loaded Hadoop catalog {}", catalog);
           return Optional.of(catalog);
         case HIVE:
-          catalog = HiveCatalogs.loadCatalog(conf);
+          catalog = CatalogUtil.loadCatalog(HiveCatalog.class.getName(), "hive", ImmutableMap.of(), conf);
```
Comment (marton-bod) [inline mr/src/main/java/org/apache/iceberg/mr/Catalogs.java:217]: Shall we replace the "hive" literals with `CatalogUtil.ICEBERG_CATALOG_TYPE_HIVE`?

File: hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java (hunk 23 above)
Comment (lcspinter) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java:39]: Good idea.

File: mr/src/main/java/org/apache/iceberg/mr/Catalogs.java (hunk 24 above)
Comment (lcspinter) [reply mr/src/main/java/org/apache/iceberg/mr/Catalogs.java]: Done.

**Review by marton-bod — APPROVED** [review]

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 16 above)
Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java]: @rdblue I made a few changes compared to the original version of the HiveClientPoolCache. Now the cleanup and update are completely delegated to the cache itself. Could you please review it? Thanks

**Comment by pvary** [comment]
@rymurr, @rdblue: Any more comments? I think this would be ready to be merged.

**Review by pvary — APPROVED** [review]

**Comment by rymurr** [comment]
Hey @pvary I am on holiday this week so no need to wait for me!

**Comment by pvary** [comment]
Merged the PR. Thanks for the PR @lcspinter and @marton-bod, @rymurr, @rdblue  for the review!

File: hive-metastore/src/main/java/org/apache/iceberg/hive/ClientPoolImpl.java [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+
+import java.io.Closeable;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class ClientPoolImpl<C, E extends Exception> implements Closeable, ClientPool<C, E> {
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/ClientPoolImpl.java:30]: @lcspinter, @pvary, why was it necessary to create an abstract client pool implementation?

Comment (lcspinter) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/ClientPoolImpl.java:30]: @rdblue The `ClientPoolImpl` is basically just the renamed copy of `ClientPool` from the previous version of the code, with a slight modification. What changed is that the `Action` interface and `run` method was extracted into a separate `ClientPool` interface (That's why we had to do the renaming). This interface is implemented by the `CacheClientPool` and `ClientPoolImpl` classes (and indirectly by `HiveClientPool`). 
When someone calls `run` on an instance of `CachedClientPool`, we just fetch the appropriate `HiveClientPool` instance from the cache and call `run` on that instance.

Comment (rdblue) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/ClientPoolImpl.java:30]: Thanks, @lcspinter! The context I was missing was that the `CacheClientPool` implements the interface.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/ClientPoolImpl.java [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+          // wake every second in case this missed the signal
+          synchronized (signal) {
+            signal.wait(1000);
+          }
+        }
+      }
+
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      LOG.warn("Interrupted while shutting down pool. Some clients may not be closed.", e);
```
Comment (StefanXiepj) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/ClientPoolImpl.java:107]: hi，why not catching InterruptedException in the while-loop and continue to close others client? or thrown exception directly?
